### PR TITLE
[API-18] New file service endpoints and workflow support

### PR DESCRIPTION
**Pull request by betacar**
Closes #54, #53, #52, #50, #47, #46, #45, and #59.

### CODE REVIEW

File: tests/src/api/index.test.js [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+      api.send('POST', '/', {
+        chunk: 0,
+        size: 100
+      }).catch(error => {
+        expect(error).toEqual(expectedResponse);
+      });
+    });
+  });
+
+  describe('with unsopported response status code', () => {
```
Comment (Arpit-Sharma-USC) [inline tests/src/api/index.test.js:266]: small typo here `unsopported` -> `unsupported`

**Review by Arpit-Sharma-USC — COMMENTED** [review]
my watch has ended!

File: tests/src/api/index.test.js (hunk 1 above)
Comment (betacar) [reply tests/src/api/index.test.js]: Good catch!

File: src/index.js [hunk 2]
```diff
@@ -587,55 +586,14 @@ export default class Bynder {
...
-   * * @param {Number} fileSize - File byte size
-   * @return {Promise}
+   * @param {String} fileId Unique file identifier
+   * @param {String} fileName Original file name
+   * @param {Number} chunksCount Number of chunks
+   * @param {Number} fileSize File byte size
+   * @return {Promise<object>} Object containing the correlation ID (`correlationId`) and file ID (`fileId`) of the upload
    */
   finaliseUpload(fileId, fileName, chunksCount, fileSize) {
     return this.api.send('POST', `v7/file_cmds/upload/${fileId}/finalise`, {
```
Comment (Arpit-Sharma-USC) [inline src/index.js:633]: Also this might need changing. This api call also needs `sha256` (as hexdigest)
`intent`. See here: https://bynder.atlassian.net/wiki/spaces/BAR/pages/776339703/File+service+interface+contract

File: tests/src/index.test.js [hunk 3]
```diff
@@ -240,19 +324,89 @@ describe('#uploadFileInChunks', () => {
...
+
+    const correlation = await bynder.finaliseUpload(fileId, file.filename, 1, file.body.length);
+    expect(correlation).toBeDefined();
+    expect(bynder.api.send).toHaveBeenNthCalledWith(1, 'POST', `v7/file_cmds/upload/${fileId}/finalise`, {
+      chunksCount: 1,
+      fileName: file.filename,
+      fileSize: file.body.length
+    });
+  });
+
```
Comment (Arpit-Sharma-USC) [inline tests/src/index.test.js:353]: Based on the above comment I think you should also alter this test case.

**Review by Arpit-Sharma-USC — COMMENTED** [review]
can you check and confirm this wrt to `finalise` behaviour ?

File: src/index.js [hunk 4]
```diff
@@ -0,0 +1,752 @@
...
+    this._chunkNumber = 0;
+
+    // Iterate over the chunks and send them
+    while (this._chunkNumber <= (this._chunks - 1)) {
+      const start = this._chunkNumber * FILE_CHUNK_SIZE;
+      const end = Math.min(start + FILE_CHUNK_SIZE, size);
+      const chunk = body.slice(start, end);
+
+      await this.api
+        .send('POST', `v7/file_cmds/upload/${fileId}/chunk/${this._chunkNumber}`, { chunk })
```
Comment (Arpit-Sharma-USC) [inline src/index.js:690]: Here FS is expecting a sha-256 to be passed for each chunk. They verify that on their end to check for corrupted chunks.

` headers["content-sha256"] = sha256(each_chunk).hexdigest` (This is the python version of this which)

Comment (Arpit-Sharma-USC) [reply src/index.js:690]: See steps 36-38 [here](https://api.media.atlassian.com/file/668ea053-fb36-4122-812b-046fb645cc37/binary?token=eyJhbGciOiJIUzI1NiJ9.eyJpc3MiOiI0MmU0MjMyMy0zYzcwLTRiYmEtYjJkMC1mYTQ0YmQ4NzAxNjIiLCJhY2Nlc3MiOnsidXJuOmZpbGVzdG9yZTpmaWxlOjY2OGVhMDUzLWZiMzYtNDEyMi04MTJiLTA0NmZiNjQ1Y2MzNyI6WyJyZWFkIl19LCJleHAiOjE2MDcwODMwODAsIm5iZiI6MTYwNzAwMDEwMH0.2Tg8bAzJaPw7dCgi65MyNvTcQX-f4c2dmgNq5WwE5Ro&client=42e42323-3c70-4bba-b2d0-fa44bd870162&name=upload-streamlined.png)

File: tests/src/api/index.test.js [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+      expect(response).toEqual(expectedResponse);
+      expect(api.axios.request).toHaveBeenNthCalledWith(1, {
+        url: '/',
+        method: 'POST',
+        data: 'chunk=0&size=100',
+        headers: {
+          'User-Agent': `bynder-js-sdk/${pkg.version}`,
+          'Authorization': 'Bearer token',
+          'Content-Type': 'application/x-www-form-urlencoded'
+        },
```
Comment (Arpit-Sharma-USC) [inline tests/src/api/index.test.js:150]: I feel this might change after you alter the header for the sha256 as well?

File: tests/src/api/index.test.js [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+        method: 'GET',
+        params: {
+          chunk: 0,
+          size: 100
+        },
+        data: null,
+        headers: {
+          'User-Agent': `bynder-js-sdk/${pkg.version}`,
+          'Authorization': 'Bearer token'
+        },
```
Comment (Arpit-Sharma-USC) [inline tests/src/api/index.test.js:210]: Same as above

**Review by Arpit-Sharma-USC — COMMENTED** [review]
can you also update the sha header

**Review by Arpit-Sharma-USC — APPROVED** [review]
crappy_checkmark

File: src/index.js [hunk 7]
```diff
@@ -608,7 +609,7 @@ export default class Bynder {
    * @return {Promise<object>} Object containing the correlation ID (`correlationId`) and file ID (`fileId`) of the upload
    */
   finaliseUpload(fileId, fileName, chunksCount, fileSize) {
-    return this.api.send('POST', `v7/file_cmds/upload/${fileId}/finalise`, {
+    return this.api.send('POST', `v7/file_cmds/upload/${fileId}/finalise_api`, {
```
Comment (Arpit-Sharma-USC) [inline src/index.js:612]: 👍

**Review by Arpit-Sharma-USC — APPROVED** [review]
kudos!

File: README.md [hunk 8]
```diff
@@ -166,15 +169,19 @@ bynder
 
 ### Upload
 
+- `prepareUpload()`
+- `uploadFileInChunks(fileObject, fileId, fileSize)`
+- `saveAsset(queryObject)`
 - `uploadFile(fileObject)`
+- `finaliseUpload(fileObject, fileName, chunks, fileSize)`
```
Comment (githuib) [inline README.md:176]: Wouldn't there still be just one `uploadFile(fileObject)` for the user? Wouldn't make sense to me to expose the internal calls like prepare/uploadChunk/etc. as well here.

File: src/utils/index.js [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+ * Returns file size
+ * @param {Object} file File
+ * @returns {Number} The amount of data that can be read from the file
+ */
+export function getLength(file) {
+  const { body, length } = file;
+  const bodyType = bodyTypes.get(body);
+
+  if (bodyType === bodyTypes.BUFFER) {
+    return body.length;
```
Comment (githuib) [inline src/utils/index.js:41]: What would `length` be in this case? And don't we always need the body length anyway for our purpose?

File: src/api/index.js [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+      body = params;
+
+      if (!isV6orV7) {
+        body = queryString.stringify(params);
+        // Some older endpoints require this
+        options.additionalHeaders = {
+          ...options.additionalHeaders,
+          'Content-Type': 'application/x-www-form-urlencoded'
+        };
+      }
```
Comment (githuib) [inline src/api/index.js:88]: I'm not quite sure if this is actually the case, that 'Content-Type': 'application/x-www-form-urlencoded' for all "old" API endpoints.

And even if it's the case, I'd rather do such a refactor in a separate pull request.

File: src/index.js [hunk 11]
```diff
@@ -0,0 +1,748 @@
...
+
+      this.sha256 = undefined;
+
+      return { fileId, correlationId, ...asset };
+    } catch (error) {
+      return Promise.reject({
+        status: error.status || 0,
+        message: error.message || `File not uploaded: ${filename}`
+      });
+    }
```
Comment (githuib) [inline src/index.js:746]: Wow this file is pretty huge. Can't we move at least the upload-related stuff to a separate file?

File: src/index.js [hunk 12]
```diff
@@ -0,0 +1,748 @@
...
+   * @return {Promise<object>} The information of the uploaded file, including IDs and all final file urls.
+   */
+  async uploadFile(file) {
+    const { body, filename, data } = file;
+    const { brandId } = data;
+    const bodyType = bodyTypes.get(body);
+    const size = getLength(file);
+
+    if (!brandId) {
+      return rejectValidation('upload', 'brandId');
```
Comment (githuib) [inline src/index.js:713]: We don't need a brandId if we save to an existing asset (so if we go this route: https://bynder.docs.apiary.io/#reference/upload-assets/6-saving-a-completed-upload/save-as-a-new-asset-version )

File: src/index.js [hunk 13]
```diff
@@ -0,0 +1,748 @@
...
+   * @see {@link https://bynder.docs.apiary.io/#reference/upload-assets/4-finalise-a-completely-uploaded-file/save-as-a-new-asset}
+   * @param {Object} data Asset data
+   * @param {String} data.brandId Brand ID
+   * @param {String} data.fileId File ID
+   * @param {String} data.mediaId Media ID
+   * @return {Promise<object>}
+   */
+  saveAsset(data) {
+    if (!data.brandId) {
+      return rejectValidation('upload', 'brandId');
```
Comment (githuib) [inline src/index.js:641]: We don't need a brandId when there is a mediaId.

File: src/index.js [hunk 14]
```diff
@@ -0,0 +1,748 @@
...
+  }
+
+  /**
+   * Saves a media asset in Bynder. If media id is specified in the data a new version of the asset will be saved.
+   * Otherwise a new asset will be saved.
+   * @see {@link https://bynder.docs.apiary.io/#reference/upload-assets/4-finalise-a-completely-uploaded-file/save-as-a-new-asset}
+   * @param {Object} data Asset data
+   * @param {String} data.brandId Brand ID
+   * @param {String} data.fileId File ID
+   * @param {String} data.mediaId Media ID
```
Comment (githuib) [inline src/index.js:636]: Passing a `name` in the data can be useful as well, since otherwise the asset will have no name.

From the api doc: "Warning: The asset will have no name if this is left empty."
https://bynder.docs.apiary.io/#reference/upload-assets/6-saving-a-completed-upload/save-as-a-new-asset

File: src/index.js (hunk 11 above)
Comment (betacar) [reply src/index.js:746]: Actually, this file was heavily reduced in size mostly due to separation of concerns (the API client, constant, utilities, and others were moved to their own context).

File: src/api/index.js (hunk 10 above)
Comment (betacar) [reply src/api/index.js:88]: This in-line with the current release, which has it for all old endpoints. That's why I'm making the distinction between new and old endpoints.

However, I'm all in favour of moving this later to specific endpoints that actually required it.

File: README.md (hunk 8 above)
Comment (betacar) [reply README.md:176]: I agree. Although the old methods have JS docs linking to our APIary docs.

However, I'll make those methods private now.

File: src/utils/index.js (hunk 9 above)
Comment (betacar) [reply src/utils/index.js:41]: Is the file size. Or, in Node terms, the length of the buffer.

File: src/index.js (hunk 14 above)
Comment (githuib) [reply src/index.js:638]: What do you think of this? In some other SDKs we just pass the filename as the asset name for example.

File: src/utils/index.js (hunk 9 above)
Comment (githuib) [reply src/utils/index.js:41]: Yes, but I mean, what's the difference with the length from `const { body, length } = file` which will be returned if this if condition is not true?

File: src/index.js (hunk 11 above)
Comment (githuib) [reply src/index.js:741]: wow again :D Still would prefer to have it split up a bit more, but fine for now

File: src/index.js [hunk 15]
```diff
@@ -0,0 +1,743 @@
...
+    if (fileId) {
+      url += `${fileId}/`;
+    }
+
+    return this.api.send('POST', url, data);
+  }
+
+  /**
+   * Uploads arbirtrarily sized buffer or stream file to provided S3 endpoint in chunks and registers each chunk to Bynder.
+   * Resolves the passed init result and final chunk number.
```
Comment (githuib) [inline src/index.js:655]: Seems like this is the old description?

File: src/index.js [hunk 16]
```diff
@@ -0,0 +1,743 @@
...
+    this._chunkNumber = 0;
+
+    // Iterate over the chunks and send them
+    while (this._chunkNumber <= (this._chunks - 1)) {
+      const start = this._chunkNumber * FILE_CHUNK_SIZE;
+      const end = Math.min(start + FILE_CHUNK_SIZE, size);
+      const chunk = body.slice(start, end);
+      const sha256 = create256HexHash(chunk);
+
+      await this.api.send('POST', `v7/file_cmds/upload/${fileId}/chunk/${this._chunkNumber}`, chunk, {
```
Comment (githuib) [inline src/index.js:679]: Is there any specific reason this call is `await`ed, but none of the other `this.api.send()` calls in this file are?

File: samples/index.js [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+    type: 'image',
+    limit: 9,
+    page: 1
+  })
+  .then(data => { // eslint-disable-line no-unused-vars
+    // TODO Handle data
+  })
+  .catch(error => { // eslint-disable-line no-unused-vars
+    // TODO Handle the error
+  });
```
Comment (githuib) [inline samples/index.js:27]: So all samples were removed and replaced with this?

File: src/api/index.js (hunk 10 above)
Comment (githuib) [reply src/api/index.js:87]: Well if we're sure this is the case for every `/api/` or `/v4/api/` request it's fine. Although in that case I think it would be a bit safer to check if the path starts with that, instead of checking if it starts with `/v6/` or `/v7/`.

File: src/index.js (hunk 14 above)
Comment (betacar) [reply src/index.js:638]: We could align it with the other SDKs, but that would mean more breaking changes for this new version. Either way I'm OK with any, or even both.

File: src/index.js (hunk 16 above)
Comment (betacar) [reply src/index.js:679]: Yup. We want to wait until each chunk is received successfully before sending the next one. This creates a short circuit mechanism that would allow us later to pause the upload, or even re-upload a specific chunk that wasn't uploaded previously.

File: samples/index.js (hunk 17 above)
Comment (betacar) [reply samples/index.js:27]: All previous samples were browser specific. Since we planned to remove browser support, this is a Node only. I agree it could be expanded tho.

File: src/index.js (hunk 15 above)
Comment (betacar) [reply src/index.js:655]: 👍🏼

File: src/index.js (hunk 14 above)
Comment (betacar) [reply src/index.js:638]: Name can be sent as well :)

Comment (githuib) [reply src/index.js:638]: Yeah but I meant it's not included in the `@param {String} ...` docstrings. So I thought that could be helpful for the user to see what relevant fields there are to set.

File: src/index.js (hunk 16 above)
Comment (githuib) [reply src/index.js:679]: Wouldn't we want this for the other requests as well?

File: samples/index.js (hunk 17 above)
Comment (githuib) [reply samples/index.js:27]: Yeah but we'll soon work on the samples anyway so fine for now I'd say.

File: src/index.js (hunk 14 above)
Comment (betacar) [reply src/index.js:638]: Is a couple of lines [below](https://github.com/Bynder/bynder-js-sdk/blob/add/file-service/src/index.js#L639).

File: src/utils/index.js (hunk 9 above)
Comment (betacar) [reply src/utils/index.js:59]: This is more of a left over of the old implementation. I'm changing things around to support stream objects.

File: src/index.js [hunk 18]
```diff
@@ -0,0 +1,743 @@
...
+   * @param {String} fileId Unique file identifier
+   * @param {String} fileName Original file name
+   * @param {Number} chunksCount Number of chunks
+   * @param {Number} fileSize File byte size
+   * @return {Promise<object>} Object containing the correlation ID (`correlationId`) and file ID (`fileId`) of the upload
+   */
+  _finaliseUpload(fileId, fileName, chunksCount, fileSize) {
+    return this.api.send('POST', `v7/file_cmds/upload/${fileId}/finalise_api`, {
+      chunksCount, fileName, fileSize,
+      intent: DEFAULT_UPLOAD_INTENT,
```
Comment (githuib) [inline src/index.js:615]: This can be left out, since it will fallback to the correct one for the SDKs (which is not `upload_main_uploader_asset`, that one is for the UI uploader)

File: src/constants/index.js [hunk 19]
```diff
@@ -0,0 +1,6 @@
+'use strict';
+
+export const DEFAULT_ASSETS_NUMBER_PER_PAGE = 50;
+export const FILE_CHUNK_SIZE = 1024 * 1024 * 5;
+export const WEBSOCKET_TOPIC_NAME = 'files-messages';
```
Comment (githuib) [inline src/constants/index.js:5]: Probably a leftover from when we wanted to do the websockets approach?

File: src/index.js [hunk 20]
```diff
@@ -676,21 +699,80 @@ export default class Bynder {
...
+    const { body } = file;
+
+    switch (bodyType) {
+    case bodyTypes.BUFFER:
+      await this._uploadBufferFile(body, fileId, size);
+      break;
+
+    case bodyTypes.STREAM:
+      this._chunks = await this._uploadStreamFile(body, fileId);
+      break;
```
Comment (githuib) [inline src/index.js:766]: Do we need to support both options? And wouldn't it be easier to just pass the file path, instead of a stream / buffer? This is how we do it in the other SDKs but I'm not sure about the history of this SDK.

Comment (betacar) [reply src/index.js:766]: In Node, Buffers are usually used for small size files. Streams are commonly used for larger ones.

Although we could ask for the path, I would prefer to leave this decision on the developer hands. That way we could avoid having the calculate file size each time.

Comment (betacar) [reply src/index.js:765]: On this SDK in particular, we have always expected a Buffer or Stream file body. Not a path.

File: src/index.js (hunk 18 above)
Comment (betacar) [reply src/index.js]: Fixed.

File: src/constants/index.js (hunk 19 above)
Comment (betacar) [reply src/constants/index.js]: Removed.

File: tests/src/index.test.js [hunk 21]
```diff
@@ -0,0 +1,2199 @@
...
+      bynder.getSmartfilters()
+        .catch(error => {
+          expect(error).toEqual({
+            status: 400,
+            message: 'There was a problem saving the asset'
+          });
+        });
+    });
+  });
+});
```
Comment (Arpit-Sharma-USC) [inline tests/src/index.test.js:2199]: This is a big ass file lol

File: .github/workflows/publish.yml [hunk 22]
```diff
@@ -13,7 +13,7 @@ jobs:
 
     strategy:
       matrix:
-        node-version: [ 8.x, 10.x, 11.x ]
+        node-version: [ 12.x, 13.x, 14.x, 15.x ]
```
Comment (Arpit-Sharma-USC) [inline .github/workflows/publish.yml:16]: I wouldn't change this just yet as this might break the release flow. Currently this is even with what we had with travis.

Changing this and `npm install` in the following comment might need to be tested for proper publishing.

It would be better to have a ticket for upgrading the node-versions and have this in a separate PR.

File: .github/workflows/publish.yml [hunk 23]
```diff
@@ -35,7 +35,7 @@ jobs:
         with:
           node-version: 12
           registry-url: https://registry.npmjs.org/
-      - run: yarn install
+      - run: npm install
```
Comment (Arpit-Sharma-USC) [inline .github/workflows/publish.yml:38]: Same as above. It would be better to have a ticket for upgrading the node-versions and have this in a separate PR after we test for publishing.

File: .github/workflows/run-tests.yml [hunk 24]
```diff
@@ -13,7 +13,7 @@ jobs:
 
     strategy:
       matrix:
-        node-version: [ 8.x, 10.x, 11.x ]
+        node-version: [ 12.x, 13.x, 14.x, 15.x ]
```
Comment (Arpit-Sharma-USC) [inline .github/workflows/run-tests.yml:16]: Same as above. It would be better to have a ticket for upgrading the node-versions and have this in a separate PR after we test for publishing.

I remember there were some dependency issues when I played around with different node versions.

File: .github/workflows/publish.yml (hunk 22 above)
Comment (betacar) [reply .github/workflows/publish.yml:16]: The thing is:
- Those old Node version are no longer under maintenance. 
- Replacing the `npm install` with `yarn install` can have multiple dependency consequences, since there's no `yarn.lock` file.

File: .github/workflows/publish.yml (hunk 23 above)
Comment (betacar) [reply .github/workflows/publish.yml:38]: Same comment as above.

File: .github/workflows/run-tests.yml (hunk 24 above)
Comment (betacar) [reply .github/workflows/run-tests.yml:16]: Same as above.

**Review by Arpit-Sharma-USC — CHANGES_REQUESTED** [review]
The FS stuff looks good to me. The only changes I would say are not merging the node-version upgrade as part of this and have a separate ticket so it can be tested for publishing and any side-effects 😄

File: .github/workflows/publish.yml (hunk 22 above)
Comment (Arpit-Sharma-USC) [reply .github/workflows/publish.yml:16]: I see that which is why I suggest we have a ticket for this and properly test any dependency issues that might arise by changing this. I remember there were dependency  issues when I changed node versions and the `yarn install `to `npm install`.

The current yml works without issues so to change it I would say we do it as part of a separate ticket an not this PR.

**Comment by Arpit-Sharma-USC** [comment]
> The FS stuff looks good to me. The only changes I would say are not merging the node-version upgrade as part of this and have a separate ticket so it can be tested for publishing and any side-effects 😄

Will not need this anymore as this PR depends on the new node-versions 👍

File: .github/workflows/publish.yml (hunk 22 above)
Comment (Arpit-Sharma-USC) [reply .github/workflows/publish.yml:16]: Resolved by testing this PR on the test npm account 👍

**Review by Arpit-Sharma-USC — APPROVED** [review]

File: src/utils/index.js (hunk 9 above)
Comment (betacar) [reply src/utils/index.js:59]: Going back to this, the `body.length` if just a fallback. If the `body` doesn't have that property, the `uploadFile` method will raise an exception. We cannot upload files that have `0` size.

Think about it as a fallback.

File: src/api/index.js (hunk 10 above)
Comment (betacar) [reply src/api/index.js]: Fixed!

File: samples/cli_app.js [hunk 25]
```diff
@@ -0,0 +1,50 @@
+const fs = require('fs');
```
Comment (githuib) [inline samples/cli_app.js:1]: I would keep this for the sample ticket (API-188) to avoid any more scope creep. As it already got a bit out of hand.

File: src/api/index.js [hunk 26]
```diff
@@ -0,0 +1,130 @@
...
+
+    if (method.toLocaleLowerCase() === 'post') {
+      body = params;
+
+      if (isV4orV5) {
+        body = queryString.stringify(params);
+        // Some older endpoints require this
+        options.additionalHeaders = {
+          ...options.additionalHeaders,
+          'Content-Type': 'application/x-www-form-urlencoded'
```
Comment (githuib) [inline src/api/index.js:85]: Will this work for requests that don't have a body (such as `/api/v4/media/{id}/save/{fileId}`)?

Wondering since in the java sdk this would result in an error in the framework used there.

File: src/utils/index.js (hunk 9 above)
Comment (githuib) [reply src/utils/index.js:59]: Wouldn't we want to (re)throw that exception instead of returning `length` (which would be 0) in that case?

**Review by githuib — COMMENTED** [review]
Allright looking good! Just some minor concerns and I would leave the cli sample out of this PR.

File: src/api/index.js (hunk 26 above)
Comment (betacar) [reply src/api/index.js:85]: Yes. As a matter of fact, axios always sends an empty body by default.

**Review by githuib — APPROVED** [review]
